src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -331,6 +331,7 @@ private void EventLoop()
                             if (context != null)
                             {
                                 context.HandleEvents(_buffer[i].Events);
+                                context = null;
```

Comment:
Is this needed?

Reply:
Ah, you mentioned this. It's worth adding a comment in the source code.

Reply:
@wfurt, was this necessary for release as well as debug?

Reply:
yes, but it is not needed to fix the leak. When I was running the test 1-2 handles would stay alive and test would fail. gcroot pointed here as we hold reference to context and that keep reference to SafeHandle until next engine loop. I was hesitant at first to add this but I saw similar behavior with the repro case. It sends batch of 1000 requests and after that few sockets would hang out for very long time. They would go away if I run another batch. I think this will help with cleaning if there are gaps in network activity. For the test, it was also passing if other tests were running in parallel. But that make the test depending on other tests. If we want to remove this, we would need to find better way how to turn crank on all engines.


this is Release without the null assignment. Most handles would go but few would remain alive.
```
 /mnt/github/wfurt-corefx2/artifacts/bin/tests/System.Net.Sockets.Tests/netcoreapp-Linux-Release-x64 /mnt/github/wfurt-corefx2/src/System.Net.Sockets/tests/FunctionalTests

     System.Net.Sockets.Tests.DisposedSocket.NonDisposedSocket_SafeHandlesCollected(clientAsync: True) [FAIL]
        Assert.Equal() Failure
        Expected: 0
        Actual:   4
        Stack Trace:
          /mnt/github/wfurt-corefx2/src/System.Net.Sockets/tests/FunctionalTests/DisposedSocketTests.netcoreapp.cs(29,0): at System.Net.Sockets.Tests.DisposedSocket.NonDisposedSocket_SafeHandlesCollected(Boolean clientAsync)
          --- End of stack trace from previous location where exception was thrown ---
      System.Net.Sockets.Tests.DisposedSocket.NonDisposedSocket_SafeHandlesCollected(clientAsync: True) [FINISHED] Time: 0.0483496s
    Finished:    System.Net.Sockets.Tests
```

Reply:
@AndyAyersMS, is there an issue here then with lifetime tracking?
https://github.com/dotnet/corefx/blob/42e11c64d878d5875dbc6f5d228cb034ee4b2e59/src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs#L328-L335
Seems like there's no reason for the JIT to extend the lifetime of the `context` local.

Reply:
The jit may extend lifetimes in some cases. Here the use of `out ... context` may complicate lifetime tracking. If you can dump the gcinfo for this method it would be interesting to see if `context` ends up untracked.

Even if this is what's happening, it would only keep one context object live.

Reply:
> Even if this is what's happening, it would only keep one context object live.

My assumption here is that @wfurt was testing on a machine large enough to have multiple (likely 4 given the results) event loops.  @wfurt, how many logical processors does your test machine have?

Reply:
This is 8 core physical machine. And yes, I think it is one context per SocketAsyncEngine. 
I was originally thinking about simply adding some extra work to test loop but multiple engines makes that more tricky.

Reply:
One per engine makes sense.

@wfurt sent me the assemblies to look at, an I can confirm my original hunch: in both debug/release, `context` is an untracked lifetime, caused by it being an “out” parameter.

That means the stack slot holding `context` is always reported to the GC, so explicit nulling is needed if you want to keep the lifetime as short as possible.

Reply:
@AndyAyersMS, is there a way it could be written that would negate the need for the `context = null`, or that's just the nature of things given the shape of `TryGetValue`?

Reply:
The typical fix is to isolate the problematic local in an inlinee, something like:

```C#
  else
  {
      Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
      HandleEventsInContext(handle);
  }


void HandleEventsInContext(IntPtr handle)
{      
      _handleToContextMap.TryGetValue(handle, out SocketAsyncContext context);
      if (context != null)
      {
          context.HandleEvents(_buffer[i].Events);
      }
}
```
You can even mark this new method with `AggressiveInlinining`; the jit does a little bit of explicit nulling on its own for inlinee locals so that inlining doesn't stretch lifetimes. That would catch this case (see dotnet/coreclr#9479).

Reply:
Ok, thanks for the info.